Compilation should check gcc version

The build process says it requires 5.4.1 but doesn't check this, which can lead to hilarity:
```
partysaurus:tock pal$ make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C boards/hail/
   Compiling hail v0.1.0 (file:///Users/pal/src/tock/boards/hail)
error: linking with `arm-none-eabi-gcc` failed: exit code: 1
  |
  = note: "arm-none-eabi-gcc" "-L" "/Users/pal/.xargo/lib/rustlib/thumbv7em-none-eabi/lib" "/Users/pal/src/tock/boards/hail/target/thumbv7em-none-eabi/release/deps/hail-fe1d81bd21ee0459.0.o" "-o" "/Users/pal/src/tock/boards/hail/target/thumbv7em-none-eabi/release/deps/hail-fe1d81bd21ee0459" "-Wl,--gc-sections" "-nodefaultlibs" "-L" "/Users/pal/src/tock/boards/hail/target/thumbv7em-none-eabi/release/deps" "-L" "/Users/pal/src/tock/boards/hail/target/release/deps" "-L" "/Users/pal/.xargo/lib/rustlib/thumbv7em-none-eabi/lib" "-Wl,-Bstatic" "/var/folders/hv/f4t2f5bs3gq8sk3n_rp728rw0000gn/T/rustc.qQqkSrv4kPBy/libcompiler_builtins-5fb302b8d1fd26d8.rlib" "-nostartfiles" "-Tlayout.ld" "-Wl,-Bdynamic"
  = note: ../kernel_layout.ld:32: nonconstant expression for length
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error(s)

error: Could not compile `hail`.

To learn more, run the command again with --verbose.
make[1]: *** [target/thumbv7em-none-eabi/release/hail] Error 101
make: *** [hail] Error 2
partysaurus:tock pal$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (GCC) 4.8.3

partysaurus:tock pal$ export PATH=/opt/arm/bin:$PATH
partysaurus:tock pal$ which arm-none-eabi-gcc
/opt/arm/bin/arm-none-eabi-gcc
partysaurus:tock pal$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.4.1 20160609 (release) [ARM/embedded-5-branch revision 237715]
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
partysaurus:tock pal$ make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C boards/hail/
   Compiling hail v0.1.0 (file:///Users/pal/src/tock/boards/hail)
    Finished release [optimized] target(s) in 7.19 secs
   text	   data	    bss	    dec	    hex	filename
  98144	   4856	  60672	 163672	  27f58	target/thumbv7em-none-eabi/release/hail
```

--- a/boards/Makefile.common
+++ b/boards/Makefile.common
@@ -12,6 +12,7 @@ MINIMUM_XARGO_VERSION ?= 0.3.8
 # This will hopefully move into Cargo.toml (or Cargo.toml.local) eventually
 RUSTFLAGS_FOR_XARGO_LINKING := "-C link-arg=-nostartfiles -C link-arg=-Tlayout.ld"
 
+CC        := $(TOOLCHAIN)-gcc
 SIZE      ?= $(TOOLCHAIN)-size
 OBJCOPY   ?= $(TOOLCHAIN)-objcopy
 OBJDUMP   ?= $(TOOLCHAIN)-objdump
@@ -31,6 +32,26 @@ VERBOSE =
 endif
 
 export TOCK_KERNEL_VERSION := $(shell git describe --always || echo notgit)
+
+
+# Check that gcc version is new enough (> 5.1) - used during linking
+CC_VERSION_MAJOR := $(shell $(TOOLCHAIN)-gcc -dumpversion | cut -d '.' -f1)
+ifeq (1,$(shell expr $(CC_VERSION_MAJOR) \>= 6))
+  # no-op
+else
+  ifneq (5,$(CC_VERSION_MAJOR))
+    $(info CC=$(CC))
+    $(info $$(CC) -dumpversion: $(shell $(CC) -dumpversion))
+    $(error Your compiler is too old. Need gcc version > 5.1)
+  endif
+    CC_VERSION_MINOR := $(shell $(CC) -dumpversion | cut -d '.' -f2)
+  ifneq (1,$(shell expr $(CC_VERSION_MINOR) \> 1))
+    $(info CC=$(CC))
+    $(info $$(CC) -dumpversion: $(shell $(CC) -dumpversion))
+    $(error Your compiler is too old. Need gcc version > 5.1)
+  endif
+endif
+
 
 # Check that rustc is the right nightly - warn not error if wrong, sometimes useful to test others
 RUSTC_VERSION_STRING := rustc 1.19.0-nightly (04145943a 2017-06-19)

